transactions/read-only/parallel.md

```diff
@@ -0,0 +1,215 @@
+|-----------------------------|---------------------------------|-----------------------------|-----------------------| 
+| handshake                   | none                            | net, main (handshake check) | yes        |
+| go_away                     | none                            | net              | yes                   |
+| time                        | none                            | net              | yes                   |
+| notice                      | none                            | net              | yes                   |
+| request                     | none                            | net, main (controller::fetch_block_by_id()) | yes   |
+| sync_request                | none                            | net, main (controller::fetch_block_by_number()) | yes |
+| packed_transaction          | chainbase, forkdb, producer, net, controller | net, producer, main | no    |
+| signed_block                | chainbase, forkdb, producer, net, controller | net, producer, main | no    |
+
+### SHiP
+SHiP receives blockchain state data, saves it to files, and sends data to nodes who request it.
+
+| Request               |  Data modified                  | Safe to read-only trx? |
+|-----------------------|---------------------------------|------------------------|
+| get_status            |  internal                       | yes                    |
+| get_blocks            |  internal                       | yes                    |
+| get_blocks_ack        |  internal                       | yes                    |
+
+## Design Decisions
+
+### Window Toggling
+To support toggling between `read` and `write` windows, configurable options `read-window-time`, `write-window-time`, and `read-only-max-queued-time` are provided. 
+- From `write` to `read`: When a new read-only transaction is queued, if the time the earliest transaction has been queued exceeds `read-only-max-queued-time`; or at the end of the `write` window and read-only transaction queue not empty.
+- From `read` to `write`: when read-only transaction queue is emptied by the read-only threads; or at the end of `read` window.
+
+In the following state diagram, `longest_queued_time = now - the time when the oldest trx was queued`, `write_window_deadline = time when write window starts + write-window-time`, and `read_window_deadline = time when read window starts + read-window-time`
+
+```mermaid
+flowchart TD
+    A(((write window))) -->|push a new trx| B[longest_queued_time < read-only-max-queued-time?]
+    B -->|yes| R(((read window)))
+    B -->|no| A
+    A --> D[write_window_deadline passed?]
+    D -->|yes| F[read-only trx queue empty?]
+    D -->|no| A
+    F -->|yes| A
+    F -->|no| R
+    
+    R --> S[read-only trx queue empty?]
+    S -->|yes| A
+    S -->|no| R
+    R --> T[read_window_deadline passed?]
+    T -->|yes| A
+    T -->|no| R  
+```
```

Comment:
I think the flowchart could be simplified to:


```mermaid
flowchart TD
    A(((write window))) -->|push a new ro trx| B[longest_queued_time > read-only-max-queued-time]
    B -->|yes| R(((read window)))
    B -->|no| A
    A --> D[write_window_deadline passed <b>AND</b> read-only trx queue not empty?]
    D -->|yes| R
    D -->|no| A
    
    R --> S[read-only trx queue empty <b>OR</b> read_window_deadline passed?]
    S -->|yes| A
    S -->|no| R
```

Reply:
The checks for read-only transaction queue and read-only window time are done independently. That's why I separate them.

Reply:
Why reflect such details in a state diagram, if it doesn't add any useful information?